File: vito/core/src/main/java/com/facebook/fresco/vito/drawable/CircularBorderBitmapDrawable.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.drawable;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.BitmapDrawable;
import androidx.annotation.Nullable;
import com.facebook.drawee.drawable.DrawableUtils;
import com.facebook.fresco.vito.options.BorderOptions;
import com.facebook.infer.annotation.Nullsafe;

@Nullsafe(Nullsafe.Mode.STRICT)
public class CircularBorderBitmapDrawable extends BitmapDrawable {

  private final Paint mBorderPaint = new Paint(Paint.ANTI_ALIAS_FLAG);

  private @Nullable BorderOptions mBorderOptions;
  private int mRadius;
  private int mAlpha = 255;

  public CircularBorderBitmapDrawable(Resources res, @Nullable Bitmap bitmap) {
    super(res, bitmap);
    mBorderPaint.setStyle(Paint.Style.STROKE);
  }

  @Override
  public void draw(Canvas canvas) {
    if (mRadius == 0) return;

    if (mBorderOptions == null || mBorderOptions.padding < 0.0f || mBorderOptions.width < 0.0f) {
      super.draw(canvas);
      return;
    }

    float widthReduction =
        mBorderOptions.scaleDownInsideBorders
            ? mBorderOptions.width + mBorderOptions.padding
            : mBorderOptions.padding;

    if (widthReduction > mRadius) return;

    float centerX = getBounds().exactCenterX();
    float centerY = getBounds().exactCenterY();

    if (widthReduction > 0.0f) {
      float scale = (mRadius - widthReduction) / mRadius;
      canvas.save();
      canvas.scale(scale, scale, centerX, centerY);
      super.draw(canvas);
      canvas.restore();
    } else {
      super.draw(canvas);
    }

    if (mBorderOptions.width > 0.0f) {
      canvas.drawCircle(centerX, centerY, mRadius - mBorderOptions.width / 2, mBorderPaint);
    }
  }

  @Override
  protected void onBoundsChange(Rect bounds) {
    super.onBoundsChange(bounds);
    mRadius = Math.min(bounds.width(), bounds.height()) / 2;
  }

  public void setBorder(@Nullable BorderOptions borderOptions) {
    if (mBorderOptions == null || !mBorderOptions.equals(borderOptions)) {
      mBorderOptions = borderOptions;
      ensureBorderPaint();
      invalidateSelf();
    }
  }

  public @Nullable BorderOptions getBorder() {
    return mBorderOptions;
  }

  @Override
  public void setAlpha(int alpha) {
    super.setAlpha(alpha);
    mAlpha = alpha;
    ensureBorderPaint();
  }

  private void ensureBorderPaint() {
    if (mBorderOptions != null) {
      mBorderPaint.setStrokeWidth(mBorderOptions.width);
      mBorderPaint.setColor(DrawableUtils.multiplyColorAlpha(mBorderOptions.color, mAlpha));
    }
  }
}

File: vito/core/src/main/java/com/facebook/fresco/vito/drawable/CircularBorderBitmapDrawable.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.drawable

import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.drawable.BitmapDrawable
import com.facebook.drawee.drawable.DrawableUtils
import com.facebook.fresco.vito.options.BorderOptions
import kotlin.math.min

class CircularBorderBitmapDrawable(
    res: Resources?,
    bitmap: Bitmap?,
    private var borderOptions: BorderOptions? = null
) : BitmapDrawable(res, bitmap) {

  private val borderPaint = Paint(Paint.ANTI_ALIAS_FLAG)

  private var radius = 0
  private var _alpha = 255

  override fun draw(canvas: Canvas) {
    if (radius == 0) return
    val border = borderOptions
    if (border == null || border.padding < 0.0f || border.width < 0.0f) {
      super.draw(canvas)
      return
    }
    val widthReduction =
        if (border.scaleDownInsideBorders) border.width + border.padding else border.padding
    if (widthReduction > radius) return
    val centerX = bounds.exactCenterX()
    val centerY = bounds.exactCenterY()
    if (widthReduction > 0.0f) {
      val scale = (radius - widthReduction) / radius
      canvas.save()
      canvas.scale(scale, scale, centerX, centerY)
      super.draw(canvas)
      canvas.restore()
    } else {
      super.draw(canvas)
    }
    if (border.width > 0.0f) {
      canvas.drawCircle(centerX, centerY, radius - border.width / 2, borderPaint)
    }
  }

  override fun onBoundsChange(bounds: Rect) {
    super.onBoundsChange(bounds)
    radius = min(bounds.width(), bounds.height()) / 2
  }

  var border: BorderOptions?
    get() = borderOptions
    set(borderOptions) {
      if (this.borderOptions == null || this.borderOptions != borderOptions) {
        this.borderOptions = borderOptions
        ensureBorderPaint()
        invalidateSelf()
      }
    }

  init {
    borderPaint.style = Paint.Style.STROKE
    ensureBorderPaint()
  }

  override fun setAlpha(alpha: Int) {
    super.setAlpha(alpha)
    _alpha = alpha
    ensureBorderPaint()
  }

  private fun ensureBorderPaint() {
    borderOptions?.let {
      borderPaint.strokeWidth = it.width
      borderPaint.color = DrawableUtils.multiplyColorAlpha(it.color, _alpha)
    }
  }
}
